File: fm_index_engine/dc3.py
import random
import time

time_comp_dic = {
    1:[],
    2:[],
    3:[],
    4:[],
    5:[],
    6:[],
    7:[],
    8:[],
    9:[],
    10:[],
    11:[],
    12:[],
    13:[],
    14:[],
    15:[]
}

def build_sa_naive(text):
    return sorted(range(len(text)), key=lambda i: text[i:])

def build_sample_indices(n):
    # start = time.time()
    idx = []
    for i in range(1, n - 3, 3):
        idx.append(i)
    idx.append(n - 3)
    for i in range(2, n - 3, 3):
        idx.append(i)
    # end = time.time()
    # time_comp_dic[1].append(end - start)
    return idx

def build_triplets(text, indices):
    # start = time.time()
    triplets = []
    for i in indices:
        triplet = (text[i], text[i+1], text[i+2], i)
        triplets.append(triplet)
    # end = time.time()
    # time_comp_dic[2].append(end - start)
    return triplets

# def counting_sort(triplets, indx):
#     start = time.time()
#     count = {}
#     for triplet in triplets:
#         key = triplet[indx]
#         if key not in count:
#             count[key] = []
#         count[key].append(triplet)
    
#     sorted_triplets = []
#     for key in sorted(count.keys()):
#         sorted_triplets.extend(count[key])
#     end = time.time()
#     time_comp_dic[3].append(end - start)
#     return sorted_triplets

def counting_sort(triplets, index):
    # start = time.time()

    max_val = max(triplet[index] for triplet in triplets)

    # Step 1: Create buckets
    count = [[] for _ in range(max_val + 1)]

    # Step 2: Distribute
    for triplet in triplets:
        key = triplet[index]
        count[key].append(triplet)

    # Step 3: Flatten
    sorted_triplets = []
    for bucket in count:
        sorted_triplets.extend(bucket)
    # end = time.time()
    # time_comp_dic[3].append(end - start)
    return sorted_triplets

# def counting_sort(triplets, index):
#     start = time.time()
#     count = [[] for _ in range(len(triplets)+20)]
#     for triplet in triplets:
#         key = triplet[index]
#         count[key].append(triplet)  
#     sorted_triplets = []
#     for bucket in count:
#         sorted_triplets.extend(bucket)
#     end = time.time()
#     time_comp_dic[3].append(end - start)
#     return sorted_triplets

def radix_sort(triplets, exp):
    # start = time.time()
    for i in range(exp, -1, -1):
        triplets = counting_sort(triplets, i)
    # end = time.time()
    # time_comp_dic[4].append(end - start)
    return triplets

def create_triplet_rank(triplets):
    # start = time.time()
    rank = {}
    current_rank = 0
    prev_triplet = None
    for triplet in triplets:
        if triplet[:3] != prev_triplet:
            current_rank += 1
            prev_triplet = triplet[:3]
        rank[tuple(triplet[:3])] = current_rank
    # end = time.time()
    # time_comp_dic[5].append(end - start)
    return rank

def build_l_1_2_unique_ranks(sorted_triplets):
    # start = time.time()
    l_1_2 = [0] * (len(sorted_triplets) - 1)
    for i, triplet in enumerate(sorted_triplets[1:]):
        l_1_2[i] = triplet[3]
    # end = time.time()
    # time_comp_dic[6].append(end - start)
    return l_1_2

def create_reduced_string(text, indices, triplet_rank):
    # start = time.time()
    reduced_string = []
    for i in indices:
        triplet = (text[i], text[i+1], text[i+2])
        reduced_string.append(triplet_rank[triplet])
    # end = time.time()
    # time_comp_dic[7].append(end - start)
    return reduced_string

def build_l_1_2(sa_reduced, idx_1_2):
    # start = time.time()
    l_1_2 = [0] * (len(sa_reduced) - 1)
    for i, pos in enumerate(sa_reduced[1:]):
        l_1_2[i] = idx_1_2[pos]
    # end = time.time()
    # time_comp_dic[8].append(end - start)
    return l_1_2

def build_r_1_2(l_1_2):
    # start = time.time()
    r_1_2 = {}
    for i, idx in enumerate(l_1_2):
        r_1_2[idx] = i
    # end = time.time()
    # time_comp_dic[9].append(end - start)
    return r_1_2

def build_mod_0_buckets(text, r_1_2):
    # start = time.time()
    mod_0_buckets = []
    for i in range(0, len(text) - 3, 3):
        first_char = text[i]
        mod_0_buckets.append((first_char, r_1_2[i + 1] + 1 if i + 1 in r_1_2 else 0, i))
    # end = time.time()
    # time_comp_dic[10].append(end - start)
    return mod_0_buckets

def create_r_0(mod_0_buckets):
    # start = time.time()
    r_0 = {}
    for rank, bucket in enumerate(mod_0_buckets):
        r_0[bucket[2]] = rank
    # end = time.time()
    # time_comp_dic[11].append(end - start)
    return r_0

def create_l_0(r_0):
    # start = time.time()
    l_0 = [0] * len(r_0)
    for pos, rank in r_0.items():
        l_0[rank] = pos
    # end = time.time()
    # time_comp_dic[12].append(end - start)
    return l_0

def is_a_less_than_b(text, i, j, rank):

    if j >= len(text) - 3:
        return False
    if i >= len(text) - 3:
        return True
    
    if i % 3 != 0 and j % 3 != 0:
        return rank[i] < rank[j]
    else:
        if text[i] != text[j]:
            return text[i] < text[j]
        else:
            return is_a_less_than_b(text, i + 1, j + 1, rank)

def merge_mod0_and_mod1_2(text, l_0, l_1_2, r_1_2):
    # start = time.time()
    merged_sa = []
    i0, i1 = 0, 0
    n0, n1 = len(l_0), len(l_1_2)
    while i0 < n0 and i1 < n1:
        pos0, pos1 = l_0[i0], l_1_2[i1]
        
        if is_a_less_than_b(text, pos0, pos1, r_1_2):
            merged_sa.append(pos0)
            i0 += 1
        else:
            merged_sa.append(pos1)
            i1 += 1
    
    while i0 < n0:
        merged_sa.append(l_0[i0])
        i0 += 1
    while i1 < n1:
        merged_sa.append(l_1_2[i1])
        i1 += 1
        
    # end = time.time()
    # time_comp_dic[13].append(end - start)

    return merged_sa


def build_sa_dc3(text):

    # Base case for small strings
    if len(text) <= 3:
        return sorted(range(len(text)), key=lambda i: text[i:])

    ################################################################################################
    # Part 1: Build the suffix array for the sample indices (1 mod 3 and 2 mod 3)
    ################################################################################################

    # Step 1: Add padding to the text
    text += [0, 0, 0]  

    # Step 2: build the sample indices
    idx_1_2 = build_sample_indices(len(text))

    # Step 3: build the triplets for the sample indices
    triplets = build_triplets(text, idx_1_2)

    # Step 4: sort the triplets using radix sort
    sorted_triplets = radix_sort(triplets, 2)

    # Step 5: create the rank for the triplets
    rank = create_triplet_rank(sorted_triplets)
    
    if len(rank) == len(sorted_triplets):
        
        # Step 6: build l_1_2 directly from the sorted triplets
        l_1_2 = build_l_1_2_unique_ranks(sorted_triplets)
    
    else:

        # Step 6: build the reduced string for the recursive call
        reduced_string = create_reduced_string(text, idx_1_2, rank)

        # Step 7: recursively build the suffix array for the reduced string
        sa_reduced = build_sa_dc3(reduced_string)

        # Step 8: build l_1_2
        l_1_2 = build_l_1_2(sa_reduced, idx_1_2)

    # Step 9: build r_1_2
    r_1_2 = build_r_1_2(l_1_2)

    print("idx_1_2: ", idx_1_2)

    print("l_1_2: ", l_1_2)

    print("r_1_2: ", r_1_2)

    ################################################################################################
    # Part 2: Sort the suffixes starting at positions 0 mod 3 using the sorted sample suffixes
    ################################################################################################
    
    # Step 10: Create first character buckets for the 0 mod 3 suffixes
    mod_0_buckets = build_mod_0_buckets(text, r_1_2)

    # Step 11: Sort the 0 mod 3 suffixes using the first character buckets and the ranks of the sample suffixes
    mod_0_buckets = radix_sort(mod_0_buckets, 1)

    # Step 12: Create r_0
    r_0 = create_r_0(mod_0_buckets)

    # Step 13: Create l_0
    l_0 = create_l_0(r_0)
    

    ################################################################################################
    # Part 3: Merge the two suffix arrays (l_0 and l_1_2) to get the final suffix array
    ################################################################################################

    # Step 14: Merge the two sorted lists (l_0 and l_1_2) to get the final suffix array
    sa = merge_mod0_and_mod1_2(text, l_0, l_1_2, r_1_2)

    return sa



if __name__ == "__main__":

    # text = [1,2,2,1,1,2,1,1]
    text = [4,2,1,0,3,3]
    sa_naive = build_sa_naive(text)
    print("Suffix Array:", sa_naive)
    sa_dc3 = build_sa_dc3(text)
    print("Suffix Array (DC3):", sa_dc3)
    print("Match:", sa_naive == sa_dc3)

 
    # Generate random text for testing
    # max_tries = 3
    # same_out = True
    # while max_tries > 0 and same_out:
    #     max_tries -= 1
    #     text = [random.randint(1, 2) for _ in range(1000000)]
        # sa_naive = build_sa_naive(text)
        # sa = build_sa_dc3(text)
        # same_out = (sa_naive == sa)
        # if not same_out:
        #     print("Text:", text)
        #     print("Naive SA:", sa_naive)
        #     print("DC3 SA:", sa)
        #     print("Match:", same_out)
        #     break

    # # save the time comparison results to a file
    # with open("dc3_time_comparison.txt", "w") as f:
    #     for key, times in time_comp_dic.items():
    #         f.write(f"Step {key}: {times}\n")
    
    # # print the mean of each step's times
    # for key, times in time_comp_dic.items():
    #     if times:
    #         mean_time = sum(times) / len(times)
    #         print(f"Step {key}: Mean Time = {mean_time:.6f} seconds, Sum: {sum(times):.6f} seconds, Count: {len(times)}")

File: fm_index_engine/dc3.cpp
#include <bits/stdc++.h>

using namespace std;


std::vector<int> build_suffix_array_naive(const std::vector<int> &s) {
    int n = s.size();

    std::vector<int> sa(n);
    for (int i = 0; i < n; i++) {
        sa[i] = i;
    }

    // Sort suffix indices
    std::sort(sa.begin(), sa.end(), [&](int a, int b) {
        int n = s.size();
        while (a < n && b < n) {
            if (s[a] != s[b])
                return s[a] < s[b];
            a++;
            b++;
        }
        return a == n; // shorter suffix comes first
    });

    return sa;
}

struct Triplet {
    int a;
    int b;
    int c;
    int index;
};

std::vector<int> build_sample_indices(int n) {

    std::vector<int> idx;

    for (int i = 1; i < n - 3; i += 3) {
        idx.push_back(i);
    }

    idx.push_back(n - 3);

    for (int i = 2; i < n - 3; i += 3) {
        idx.push_back(i);
    }

    return idx;
}

std::vector<Triplet> build_triplets(const std::vector<int> &text,  const std::vector<int> &indices) {

    std::vector<Triplet> triplets;
    triplets.reserve(indices.size());

    for (int i : indices) {
        Triplet t;
        t.a = text[i];
        t.b = text[i + 1];
        t.c = text[i + 2];
        t.index = i;

        triplets.push_back(t);
    }
    
    return triplets;
}

std::vector<Triplet> counting_sort(std::vector<Triplet> &triplets, int keyIndex) {

    // Step 1: find max value for the chosen key
    int max_val = 0;

    for (const auto &t : triplets) {
        int key;
        if (keyIndex == 0) key = t.a;
        else if (keyIndex == 1) key = t.b;
        else key = t.c;

        max_val = std::max(max_val, key);
    }

    // Step 2: create buckets
    std::vector<std::vector<Triplet>> buckets(max_val + 1);

    // Step 3: distribute
    for (const auto &t : triplets) {
        int key;
        if (keyIndex == 0) key = t.a;
        else if (keyIndex == 1) key = t.b;
        else key = t.c;

        buckets[key].push_back(t);
    }
    
    // Step 4: flatten
    std::vector<Triplet> sorted_triplets;
    sorted_triplets.reserve(triplets.size());

    for (const auto &bucket : buckets) {
        for (const auto &t : bucket) {
            sorted_triplets.push_back(t);
        }
    }

    return sorted_triplets;
}

void radix_sort(std::vector<Triplet> &triplets, int exp) {
    for (int i = exp; i >= 0; i--) {
        triplets = counting_sort(triplets, i);
    }
}

std::vector<int> build_r_1_2(std::vector<int> &reduced_sa, vector<int> &indices, int n){

    std::vector<int> r_1_2(n);

    return {};
}

std::vector<int> merge_group(std::vector<int> &l_0, std::vector<int> &l_1_2, std::vector<int> &r_1_2, std::vector<int> &text) {

    int i0 = 0, i1 = 1;
    int n0 = l_0.size(), n1 = l_1_2.size();

    std::vector<int> result;
    result.reserve(n0 + n1);

    while (i0 < n0 && i1 < n1) {
        int pos0 = l_0[i0];
        int pos12 = l_1_2[i1];

        bool take0;

        if (pos12 % 3 == 1) {
            // Compare (text[pos0], rank[pos0+1]) vs (text[pos12], rank[pos12+1])
            if (text[pos0] != text[pos12]) {
                take0 = text[pos0] < text[pos12];
            } else {
                take0 = r_1_2[pos0 + 1] < r_1_2[pos12 + 1];
            }
        } else {
            // pos12 % 3 == 2
            // Compare (text[pos0], text[pos0+1], rank[pos0+2])
            // vs     (text[pos12], text[pos12+1], rank[pos12+2])
            if (text[pos0] != text[pos12]) {
                take0 = text[pos0] < text[pos12];
            } else if (text[pos0 + 1] != text[pos12 + 1]) {
                take0 = text[pos0 + 1] < text[pos12 + 1];
            } else {
                take0 = r_1_2[pos0 + 2] < r_1_2[pos12 + 2];
            }
        }

        if (take0) {
            result.push_back(pos0);
            i0++;
        } else {
            result.push_back(pos12);
            i1++;
        }
    }

    // Append remaining
    while (i0 < n0) result.push_back(l_0[i0++]);
    while (i1 < n1) result.push_back(l_1_2[i1++]);

    return result;
}

std::vector<int> build_suffix_array(const std::vector<int> &str) {

    // Base case if length of text is 3 or less built it using naive approach
    if(str.size() <= 3){
        return build_suffix_array_naive(str);
    }

    // Stack 3 zeros at the end of the text to avoid out-of-bounds when creating triplets
    std::vector<int> text = str;
    text.resize(text.size() + 3, 0);

    int n = text.size();

    std::vector<int> indices = build_sample_indices(n);
    
    std::vector<Triplet> triplets = build_triplets(text, indices);
    radix_sort(triplets, 2);

    std::vector<int> rank(n);
    int current_rank = 0;
    rank[triplets[0].index] = current_rank;
    for (int i = 1; i < triplets.size(); i++) {
        const auto &t = triplets[i];
        const auto &prev = triplets[i - 1];

        if (t.a != prev.a || t.b != prev.b || t.c != prev.c) {
            current_rank++;
        }
        rank[t.index] = current_rank;
    }

    vector<int> l_1_2(indices.size());
    if (current_rank < triplets.size()) {
        std::vector<int> reduced_text(indices.size());
        int i = 0;
        for(int idx: indices) {
            reduced_text[i] = rank[idx];
            i++;
        }
        std::vector<int> reduced_sa = build_suffix_array(reduced_text);
        
        for (i = 0; i < reduced_sa.size(); i++) {
            l_1_2[i] = indices[reduced_sa[i]];
        }

    } else {
        for (int i = 0; i < triplets.size(); i++) {
            l_1_2[i] = triplets[i].index;
        }
    }

    std::vector<int> r_1_2(n);
    for(int i = 0; i < l_1_2.size(); i++){
        r_1_2[l_1_2[i]] = i;
    }

    // Part 2
    std::vector<Triplet> buckets((n - 1) / 3);
    
    for(int i = 0; i < n - 3; i += 3){
        Triplet t;
        t.a = text[i];
        t.b = r_1_2[i + 1];
        t.index = i;

        buckets[i / 3] = t;
    }

    
    radix_sort(buckets, 1);    

    std::vector<int> l_0(buckets.size());

    for(int i = 0; i < l_0.size(); i++){
        l_0[i] = buckets[i].index;
    }
    
    std::vector<int> r_0(n - 3);
    for(int i = 0; i < l_0.size(); i++){
        r_0[l_0[i]] = i;
    }

    // Part 3 merge
    std::vector<int> merged_sa(n - 3);

    merged_sa = merge_group(l_0, l_1_2, r_1_2, text);

    return merged_sa;
}

std::vector<int> generate_random_string(const std::vector<int>& input, int n) {
    static std::mt19937 gen(std::random_device{}());  // seeded once

    std::vector<int> result;
    result.reserve(n);

    std::uniform_int_distribution<> dist(0, input.size() - 1);

    for (int i = 0; i < n; ++i) {
        result.push_back(input[dist(gen)]);
    }

    return result;
}

vector<int> build_sa_nlogn(vector<int> &text) {
    int n = text.size();

    vector<int> sa(n), rank(n), tmp(n);

    // Initial ranking based on values
    for (int i = 0; i < n; i++) {
        sa[i] = i;
        rank[i] = text[i];
    }

    for (int k = 1; k < n; k <<= 1) {
        // Custom comparator
        auto cmp = [&](int i, int j) {
            if (rank[i] != rank[j])
                return rank[i] < rank[j];

            int ri = (i + k < n) ? rank[i + k] : -1;
            int rj = (j + k < n) ? rank[j + k] : -1;
            return ri < rj;
        };

        sort(sa.begin(), sa.end(), cmp);

        // Recompute temporary ranks
        tmp[sa[0]] = 0;
        for (int i = 1; i < n; i++) {
            tmp[sa[i]] = tmp[sa[i - 1]] + (cmp(sa[i - 1], sa[i]) ? 1 : 0);
        }

        // Copy back
        rank = tmp;

        // Optimization: stop early if all ranks are unique
        if (rank[sa[n - 1]] == n - 1)
            break;
    }

    return sa;
}

int main() {

    // std::vector<int> text = {1,2,1,1,2,2,1};
    // std::vector<int> sa = build_suffix_array_naive(text);
    std::vector<int> text = generate_random_string({1, 2, 3, 4, 5, 6, 7, 8}, 10000000);
    
    double total = 0;
    int runs = 1;
    
    for (int i = 0; i < runs; ++i) {
        auto start = std::chrono::high_resolution_clock::now();

        auto sa = build_suffix_array(text);
        // auto sa = build_sa_nlogn(text);

        auto end = std::chrono::high_resolution_clock::now();

        total += std::chrono::duration<double, std::milli>(end - start).count();
    }

    std::cout << "Average time: " << total / runs << " ms\n";

    return 0;
}